### PR TITLE
CORGI-616 brew tag relation timeliness

### PR DESCRIPTION
**Pull request by jasinner**
This saves brew_tag relations everytime we ingest a new build. It's done before save_product_taxonomy so that the new build's components are instantly related to any product streams via it's brew_tags.

I wanted to make use of the existing save_product test function from test_manifest.py so I made it a test fixture. Since that function is called by multiple other functions in test_manifests, I also made all those fixtures as well. One of the fixtures (setup_products_and_components_provides) originally took arguments so I updated it to use pytest.markers instead.

### CODE REVIEW

**Comment by jasinner** [comment]
@RedHatProductSecurity/corgi-devs please review

File: corgi/tasks/common.py [hunk 1]
```diff
@@ -69,7 +69,12 @@ def get_last_success_for_task(task_name: str) -> datetime:
 
 
 def create_relations(
-    build_ids, build_type, external_system_id, product_ref, relation_type, refresh_task
+    build_ids: tuple,
+    build_type: SoftwareBuild.Type,
+    external_system_id: str,
+    product_ref: str,
+    relation_type: ProductComponentRelation.Type,
+    refresh_task,
```
Comment (juspence) [inline corgi/tasks/common.py]: refresh_task can be an Optional[Callable] I think.

File: corgi/tasks/brew.py [hunk 2]
```diff
@@ -488,34 +494,55 @@ def save_node(
...
 
 
-@app.task(
-    base=Singleton,
-    autoretry_for=RETRYABLE_ERRORS,
-    retry_kwargs=RETRY_KWARGS,
-    soft_time_limit=settings.CELERY_LONGEST_SOFT_TIME_LIMIT,
-)
-def load_brew_tags() -> None:
+def load_stream_brew_tags() -> None:
```
Comment (juspence) [inline corgi/tasks/brew.py:505]: This is still being run daily as a Celery task, but you've removed the decorator.

File: tests/test_manifests.py [hunk 3]
```diff
@@ -234,11 +223,10 @@ def test_product_manifest_excludes_unreleased_components():
     }
 
 
-def test_product_manifest_excludes_internal_components():
+@pytest.mark.internal_component(True)
```
Comment (juspence) [inline tests/test_manifests.py]: Instead of having a pytest fixture that takes arguments based on a test marker, can we just have the fixture create all the components (normal, internal, unreleased) every time?

That would also make the manifest tests more robust (by including more data), so we might catch more bugs (like the duplicate provides / upstreams bugs that we had to fix).

This test should work the same either way - we can ignore the normal and unreleased components, use the internal_provided and internal_dev_provided components, and continue to ```assert "redhat.com/" not in package["name"]```.

File: tests/test_manifests.py [hunk 4]
```diff
@@ -198,11 +188,10 @@ def test_slim_rpm_in_containers_manifest():
     assert manifest["relationships"][-1] == document_describes_product
 
 
-def test_product_manifest_excludes_unreleased_components():
+@pytest.mark.unreleased(True)
```
Comment (juspence) [inline tests/test_manifests.py]: As below - let's create all the components, ignore the normal and internal ones, and use only the unreleased components in this test. Then we can ditch the marker and the parameter in the fixture.

File: pyproject.toml [hunk 5]
```diff
@@ -21,6 +21,8 @@ markers = [
     "integration",
     "performance",
     "unit",
+    "unreleased",
+    "internal_component"
```
Comment (juspence) [inline pyproject.toml]: Using pytest marks as parameters to a test fixture seems very odd. It also means the list of marks will grow a lot if we keep adding parameters.

I think the simplest solution is just to remove the @pytest.fixture() decorator, import the setup function in each test module that needs to use them, and call the function with whatever arguments it needs, like we were doing before.

Or you could split each parameter (internal, unreleased, or regular components) into its own fixture. Each fixture can just call the main setup function with the right set of parameters for that fixture - for example, the setup_internal fixture would call the setup_base() function with an internal=True argument. This is the "fixture factories" approach described here: http://stackoverflow.com/questions/63425998/#63431123

Or we could combine the manifest tests, then run them with pytest.mark.parametrize and pass different test data to them each time. Then the helper function can remain a function, and we can pass different test data by just calling the function with different parameters.

**Review by juspence — COMMENTED** [review]
I haven't finished reviewing the Brew tags logic or the tests, but going to submit the rest before I forget.

File: corgi/tasks/brew.py [hunk 6]
```diff
@@ -488,34 +494,55 @@ def save_node(
...
-                ps.name,
-                ProductComponentRelation.Type.BREW_TAG,
-                refresh_task,
-            )
-            logger.info("Saving %s new builds for %s", no_of_created, brew_tag)
+            _create_relations_for_tag(builds, brew_tag, build_type, ps.name, refresh_task)
+
+
+def load_brew_tags(build_id: str, brew_tags: list[str]) -> int:
+    all_stream_tags = ProductStream.objects.exclude(brew_tags__exact={}).values_list(
```
Comment (juspence) [inline corgi/tasks/brew.py:514]: Should this be ```brew_tags=[]``` with square brackets to match against an empty list of tags? ```{}``` is Python syntax for an empty set, and I'm not sure that will translate properly in the JSONField.

Or maybe this JSON blob for the Brew tags is always wrapped in an outermost / top-level curly {} braces instead of square [] brackets, even when empty?

File: corgi/tasks/brew.py [hunk 7]
```diff
@@ -488,34 +494,55 @@ def save_node(
...
+    )
+    no_created = 0
+    for stream_name, stream_tags in all_stream_tags:
+        for tag in brew_tags:
+            if tag in stream_tags:
+                brew, build_type, _ = _relation_context_for_stream(stream_name)
+                logger.info(f"Creating relations for {stream_name} and {tag}")
+                no_created += _create_relations_for_tag(
+                    (build_id,), tag, build_type, stream_name, None
+                )
```
Comment (juspence) [inline corgi/tasks/brew.py]: Instead of many nested loops that check all the tags, I think it's more efficient to convert the tags to a set and check the intersection:
```suggestion
    brew_tags = set(brew_tags)
    for stream_name, stream_tags in all_stream_tags:
        build_stream_tags = set(stream_tags)
        build_stream_tags = brew_tags.intersection(build_stream_tags)
        for tag in build_stream_tags:
                brew, build_type, _ = _relation_context_for_stream(stream_name)
                logger.info(f"Creating relations for {stream_name} and {tag}")
                no_created += _create_relations_for_tag(
                    (build_id,), tag, build_type, stream_name, None
                )
```

File: corgi/tasks/brew.py [hunk 8]
```diff
@@ -113,8 +113,14 @@ def slow_fetch_brew_build(
...
         save_component(child_component, root_node, softwarebuild)
 
-    # We don't call save_product_taxonomy by default to allow async call of slow_load_errata task
-    # See CORGI-21
+    # for builds with any tag, check if the tag is used for product stream relations, and create the
+    # relations if so.
+    if not build_meta["tags"]:
+        logger.info("no brew tags")
+    else:
+        load_brew_tags(build_id, build_meta["tags"])
```
Comment (juspence) [inline corgi/tasks/brew.py]: We return no_created from this function, but it's ignored here - did you want to log it somewhere?

File: tests/test_brew_collector.py [hunk 9]
```diff
@@ -844,22 +848,45 @@ def test_fetch_rpm_build(mock_sca, mock_brew):
...
     image_index = Component.objects.get(
         name="subctl-container", type=Component.Type.CONTAINER_IMAGE, arch="noarch"
     )
-    softwarebuild = SoftwareBuild.objects.get(build_id=1781353, build_type=SoftwareBuild.Type.BREW)
 
+    # Check that new components are related to the build via the brew_tag
+    assert ProductComponentRelation.objects.get(
+        build_id="1781353", build_type=SoftwareBuild.Type.BREW
+    )
+    assert image_index.productstreams.get(pk=stream.uuid)
```
Comment (juspence) [inline tests/test_brew_collector.py]: Minor nit: It's more efficient to just ```assert Model.filter(field=condition).exists()``` if we're not going to use the returned value.

Or you can just say ```Model.get(field=condition)``` since that will raise an exception anyway if the object doesn't exist - no need for the assert here.

File: tests/test_brew_collector.py [hunk 10]
```diff
@@ -893,6 +920,34 @@ def test_fetch_container_build_rpms(mock_fetch_brew_build, mock_load_errata, moc
...
+        type=ProductComponentRelation.Type.BREW_TAG,
+    )
+    assert new_brew_tag_relation.product_ref == stream.name
+    mock_fetch_modular_build.assert_called_once()
+
+
+@pytest.mark.django_db
+def test_load_brew_tags():
+    stream = ProductStreamFactory(brew_tags={"rhacm-2.4-rhel-8-container-released": True})
+    load_brew_tags("1", ["rhacm-2.4-rhel-8-container-released"])
```
Comment (juspence) [inline tests/test_brew_collector.py:944]: It might be worth patching _create_relations_for _tag so we can assert it's called with a None argument for refresh_task.

Or we could mock both slow_fetch_brew_build and slow_fetch_modular_build, then assert_not_called() on both mocks.

**Review by juspence — COMMENTED** [review]
I think we should change how we reuse the helper methods in the tests, and we're missing the Celery task decorator now. But the rest of the Brew logic is good, I think.

File: corgi/tasks/common.py (hunk 1 above)
Comment (jasinner) [reply corgi/tasks/common.py]: I tried that, but unfortunately Callable doesn't have the '.delay()' function used below.

File: corgi/tasks/brew.py (hunk 2 above)
Comment (jasinner) [reply corgi/tasks/brew.py:505]: Oh yes, I'll add it back, thanks.

File: corgi/tasks/brew.py (hunk 7 above)
Comment (jasinner) [reply corgi/tasks/brew.py]: Good idea. I'll change it.

File: corgi/tasks/brew.py (hunk 6 above)
Comment (jasinner) [reply corgi/tasks/brew.py:514]: brew_tags attribute of Product Stream is a dictionary, not a list.

File: corgi/tasks/brew.py (hunk 8 above)
Comment (jasinner) [reply corgi/tasks/brew.py]: It's a good idea. We already log each time we create a relation, but I guess it's also a good idea to provide this count here so we know where the new relations came from.

File: tests/test_brew_collector.py (hunk 9 above)
Comment (jasinner) [reply tests/test_brew_collector.py]: Ok, I'll change it to .filter().exists() because it's more explicit what's be testing than throwing an exception in my opinion.

File: tests/test_brew_collector.py (hunk 10 above)
Comment (jasinner) [reply tests/test_brew_collector.py:944]: Ok, I like the 2nd approach, as we need the _create_relations_for_tag function to actually make the data changes in order to assert those changes are as we expect.

File: corgi/tasks/common.py (hunk 1 above)
Comment (jasinner) [reply corgi/tasks/common.py]: Actually I can use celery.local.PromiseProxy as the type instead of Callable.

File: tests/test_manifests.py (hunk 3 above)
Comment (jasinner) [reply tests/test_manifests.py]: Ok, I just moved the save_product function used by test_brew_collector.py::test_fetch_container_build_rpms into conftest.py, and leave test_manifests.py as it was (without fixtures).

File: tests/test_manifests.py (hunk 4 above)
Comment (jasinner) [reply tests/test_manifests.py]: I removed the markers.

File: pyproject.toml (hunk 5 above)
Comment (jasinner) [reply pyproject.toml]: I removed the markers.
